### ISSUE TITLE
[Epic] Pytest Improvements

### ISSUE THREAD

**Issue by ThomasLaPiana** [ethyca/fides#2135]
## Objective

The goal of this Epic is to refactor our backend automated testing, specifically our `pytest` suite, to a point where we can both guarantee quality now as well as make that level of quality maintainable into the future.

## Links

* [Original Confluence Design Doc](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2571370641/2022-12-20+-+Pytest+Automation+Refactor)
* [Code Climate](https://codeclimate.com/quality) (tool for tracking testing coverage, free for OSS)

## Issues (in order of recommended execution)
* [x] #2179
* [x] #2174
* [ ] #2175
* [x] #2176
* [ ] #2177
* [ ] #2178
* [x] #2180
* [x] #2181
* [x] #2182
* [x] #2183

**Comment by sanders41** [comment]
I did a little looking into why tests are running slow. The slowest unit test is currently `test_get_filtered_users` which takes ~ 38 seconds. I ran a cProfile on the tests and a couple things stand out to me. First it looks like in general the collection/setup is relatively slow so we could probably make improvements there that would benefit all tests.

For this specific test it looks like the slowest part is acquiring thread locks for post requests. My current theory here (but unproven) is the lock is acquired because the endpoint is synchronous which FastAPI runs in a thread pool requiring the lock. Maybe if we make the endpoint async and use the async db connection this would remove the need for a lock and speed things up?

I'm attaching the cProfile graph. I'm not sure if it will be zoomable here or not, if not and anyone wants it I can send it in slack and/or show you how I created it. 

![output](https://user-images.githubusercontent.com/25045024/211554145-a999cb15-1cc1-4937-a4ed-13898e72e41f.png)

**Comment by sanders41** [comment]
I have been thinking about our xdist discussion and the database issues around that. One thing we could explore is adding new `database` mark for tests that access the database, then run them in their own job without xdist while using xdist for all other tests. This would allow us to run at least some of the tests in parallel, I'm just not sure yet if it would be a tiny percentage of the tests and therefore not make much difference.

**Comment by ThomasLaPiana** [comment]
@sanders41 thanks for doing that analysis!

I think we want everything to be async, right? Any reason not to? I remember it being some work to get the `ctl` stuff all async but now that we have that done I don't think it would be too hard to port that to the other endpoints? If so, it sounds like we could fix our application performance and get faster tests for free! 

As for marks, I'd suspect that anything that doesn't touch the database runs extremely quickly and therefore isn't much of a concern...I'm looking at cleaning up the marks and am trying to figure out ways to speed things up. For instance, we don't need to run the database setup/cleanup before/after tests that don't touch the API/database. So i'm looking at ways to not run certain fixtures in those cases but I'm not sure how successful I'll be... i.e. `if pytest.mark.unit don't run setup_db`

**Comment by sanders41** [comment]
I think async is the way to go. We moved some ops endpoints to async for logging and the biggest issue we had there was celery doesn’t like async functions.  We had to come up with a hacky workaround for that. Between what was already done for ctl and what we found for ops hopefully we have everything we need to make the move. 

I agree about the non database test being fast and not worth xdist alone. For the database cleanup fixtures, and ones like them, one option might be to inspect the fixtures that are used for ones like `db` and fail the test with a message to include the tear down fixture if not present rather than auto using it on all tests. I don’t think any of this would make much difference with the session scoped fixtures but could make big differences with function scoped, and a lot of these are functioned scoped by necessity. The down side to this is it could get annoying to have tests fail because you forgot to include a fixture rather than having pytest automatically do it. If you think it’s something we should try I have some ideas how we might be able to do it

**Comment by sanders41** [comment]
Another thought I just had about the fixtures, it may not always be obvious if a test is accessing the database so we will have to keep that in mind. For example if the database access happens in a route, the test itself doesn’t have to have any reference to the database or a database fixture.

**Comment by sanders41** [comment]
Maybe the simple non over engineered solution here is if a test needs to access the database you have to include the fixture yourself and we turn off `autouse`. Then only tests that actually need the database get the fixture and we don't have to do any magic to figure out if it is needed.

**Comment by ThomasLaPiana** [comment]
> Maybe the simple non over engineered solution here is if a test needs to access the database you have to include the fixture yourself and we turn off `autouse`. Then only tests that actually need the database get the fixture and we don't have to do any magic to figure out if it is needed.

This is the preferable solution to me. Pytest does have some spooky (read: implicit) behavior to me in terms of fixtures and I'd like to minimize that by making our fixture usage as explicit as possible, even though it will cause a lot of added lines of code initially.
